File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        st = set()
        n = len(s)
        i = 0
        j = 0
        ans = 0
        while j < n: 
            if s[j] in st:
                while i < j and s[j] != s[i]:
                    st.remove(s[i])
                    i = i + 1
                     
                i = i + 1
                j = j + 1
                
            else:
                st.add(s[j])
                ans = max(ans, j - i + 1)
                j = j + 1
                
        return ans
        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> pre(n + 2), suf(n + 2);
        for(int i = 0; i < n; i++) {
            if(!i) {
                pre[i] = nums[i];
            }
            else {
                pre[i] = pre[i-1] * nums[i];
            }
        }
         for(int i = n-1; i >= 0; i--) {
            if(i == n-1) {
                suf[i] = nums[i];
            }
            else {
                suf[i] = suf[i+1] * nums[i];
            }
        }
        vector<int> ans(n);
        for(int i= 0; i < n; i++) {
            if(!i) {
                ans[i] = suf[i + 1];
            }
            else if(i == n-1) {
                ans[i] = pre[i-1];
            }
            else {
                ans[i] = pre[i-1] * suf[i + 1];
            }
        }
        return ans;


    }
};